[Form][TwigBridge] Don't render _method in form_rest() for a child form

fmarchalemisys opened the pull request:
The hidden `_method` must only be generated if the form is the top most form.

Always generating the hidden `_method` breaks forms using the POST method when they have children using the PUT method. If `_method` is generated for such a child form, it overrides the parent method and the form fails to validate.

See issue #23254

| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14261
| License       | MIT
| Doc PR        |

fmarchalemisys committed:
[Form][TwigBridge] Don't render _method in form_rest() for a child form

Always generating the hidden _method breaks forms using the POST method when they have children using the PUT method. The _method generated for a child form overrides the parent method and the form fails to validate.

The hidden _method must only be generated if the form is the top most form.

See issue #23254

xabbuh commented on src/Symfony/Bridge/Twig/Resources/views/Form/form_div_layout.html.twig:
```diff
@@ -303,7 +303,7 @@
     {%- endfor %}
 
-    {% if not form.methodRendered %}
+    {% if not form.methodRendered and form.parent is empty %}
```
Maybe use `is null` instead?

xabbuh commented:
ping @HeahDude @ogizanagi

ogizanagi commented:
This looks good to me appart from @xabbuh's minor suggestion which I think should be fixed despite `form_widget_compound`. Congrats on your first PR @fmarchalemisys !

fmarchalemisys committed:
[Form][TwigBridge] Use null to check for top most form

As requested, use "is null" instead of "is empty" to check if the form is a child or the top parent form. Only the top most form must output the hidden "_method".

fmarchalemisys commented:
I made the requested change i.e. replace "is empty" with "is null".

BTW, I ran the test suite against Symfony 3.3.5, not against Symfony 2.7 which is the target of this PR. Should I change the "Test pass?" entry in the PR?

xabbuh commented:
Thanks for your nice first contribution! And no, changing that line is not necessary. As you can see tests on AppVeyor (for Windows) and Travis CI (testing different PHP and dependency versions) all pass. So that's still true. :)

HeahDude commented on src/Symfony/Bridge/Twig/Resources/views/Form/form_div_layout.html.twig:
```diff
@@ -303,7 +303,7 @@
     {%- endfor %}
 
-    {% if not form.methodRendered %}
+    {% if not form.methodRendered and form.parent is null %}
```
It might break, if a field is named "parent". As discussed in https://github.com/symfony/symfony/pull/19492, I think we should be able to hardcode:
```twig
{% if not form.methodRendered and not form.hasParent %}
```
to know if a form view is a root.

ogizanagi replied:
I hesitated suggesting to add a `FormView::needsMethodToBeRendered` instead, which will do this check inside, as rendering the method input does only makes sense for a root form anyway and would avoid this issue.

Besides that, we already use `form.parent` in multiple places, so I don't think the related issue should be considered here.

HeahDude replied:
Agreed, your solution sounds good though.

ogizanagi replied:
I'll suggest it in another PR then.

ogizanagi commented:
Thanks for fixing this bug @fmarchalemisys.

ogizanagi commented:
(PR is merged, but was not closed automatically)
